### ISSUE TITLE
sway runs very slowly in qemu virto vga mode, but weston is normal.

### ISSUE THREAD

**Issue by Xiange** [swaywm/sway#7644]
Sway Version: 1.8.1

DRM info when kernel boot:
    0.217708] ACPI: bus type drm_connector registered
[    0.767585] systemd[1]: Starting Load Kernel Module drm...
[    0.784040] systemd[1]: modprobe@drm.service: Deactivated successfully.
[    0.784125] systemd[1]: Finished Load Kernel Module drm.
[    1.445949] [drm] pci: virtio-vga detected at 0000:00:02.0
[    1.470255] [drm] features: -virgl +edid -resource_blob -host_visible
[    1.470256] [drm] features: -context_init
[    1.470790] [drm] number of scanouts: 1
[    1.470795] [drm] number of cap sets: 0
[    1.470977] [drm] Initialized virtio_gpu 0.1.0 0 for 0000:00:02.0 on minor 0
[    1.472884] fbcon: virtio_gpudrmfb (fb0) is primary device
[    1.472943] virtio-pci 0000:00:02.0: [drm] drm_plane_enable_fb_damage_clips() not called
[    1.481977] virtio-pci 0000:00:02.0: [drm] fb0: virtio_gpudrmfb frame buffer device

Sway boot log:

00:00:00.000 [INFO] [sway/main.c:338] Sway version 1.8.1
00:00:00.002 [INFO] [sway/main.c:339] wlroots version 0.16.2
00:00:00.004 [INFO] [sway/main.c:120] Linux Xg64 6.3.7 #1 SMP PREEMPT_DYNAMIC Tue Jun 13 17:47:41 CST 2023 x86_64 GNU/Linux
00:00:00.005 [INFO] [sway/main.c:136] Contents of /etc/lsb-release:
00:00:00.005 [INFO] [sway/main.c:120] DISTRIB_ID="Xiange Linux x86_64"
00:00:00.005 [INFO] [sway/main.c:120] DISTRIB_RELEASE="2020-11-systemd"
00:00:00.005 [INFO] [sway/main.c:120] DISTRIB_CODENAME="Sword"
00:00:00.005 [INFO] [sway/main.c:120] DISTRIB_DISCRIPTION="Xiange Linux x86_64 with systemd"
00:00:00.005 [INFO] [sway/main.c:120]
00:00:00.005 [INFO] [sway/main.c:136] Contents of /etc/os-release:
00:00:00.005 [INFO] [sway/main.c:120] NAME="Xiange Linux"
00:00:00.005 [INFO] [sway/main.c:120] VERSION="X86_64-2020.11.11"
00:00:00.005 [INFO] [sway/main.c:120] ID=xiange
00:00:00.005 [INFO] [sway/main.c:120] PRETTY_NAME="Xiange-Linux-LiveUSB-x86_64-2020.11.11"
00:00:00.005 [INFO] [sway/main.c:120] ANSI_COLOR="1;34"
00:00:00.005 [INFO] [sway/main.c:120] HOME_URL="https://sourceforge.net/projects/xiangelinux"
00:00:00.005 [INFO] [sway/main.c:120] BUILD_ID="LiveUSB-4G-x86_64-2020.11.11"
00:00:00.005 [INFO] [sway/main.c:120]
00:00:00.005 [INFO] [sway/main.c:108] LD_LIBRARY_PATH=
00:00:00.005 [INFO] [sway/main.c:108] LD_PRELOAD=
00:00:00.005 [INFO] [sway/main.c:108] PATH=/bin:/sbin:/usr/bin:/usr/sbin
00:00:00.005 [INFO] [sway/main.c:108] SWAYSOCK=
00:00:00.005 [INFO] [sway/main.c:376] Starting sway version 1.8.1
00:00:00.009 [DEBUG] [sway/server.c:67] Initializing Wayland server
00:00:00.014 [INFO] [wlr] [libseat] [libseat/libseat.c:73] Seat opened with backend 'logind'
00:00:00.015 [INFO] [wlr] [backend/session/session.c:109] Successfully loaded libseat session
00:00:00.016 [INFO] [wlr] [backend/backend.c:220] Found 1 GPUs
00:00:00.016 [INFO] [wlr] [backend/drm/backend.c:200] Initializing DRM backend for /dev/dri/card0 (virtio_gpu)
00:00:00.016 [DEBUG] [wlr] [backend/drm/drm.c:88] Using atomic DRM interface
00:00:00.016 [DEBUG] [wlr] [backend/drm/drm.c:100] ADDFB2 modifiers unsupported
00:00:00.016 [INFO] [wlr] [backend/drm/drm.c:253] Found 1 DRM CRTCs
00:00:00.016 [INFO] [wlr] [backend/drm/drm.c:180] Found 2 DRM planes
00:00:00.016 [INFO] [wlr] [render/egl.c:201] Supported EGL client extensions: EGL_EXT_client_extensions EGL_EXT_device_base EGL_EXT_device_enumeration EGL_EXT_device_query EGL_EXT_platform_base EGL_KHR_client_get_all_proc_addresses EGL_KHR_debug EGL_EXT_platform_device EGL_EXT_platform_wayland EGL_KHR_platform_wayland EGL_EXT_platform_x11 EGL_KHR_platform_x11 EGL_EXT_platform_xcb EGL_MESA_platform_gbm EGL_KHR_platform_gbm EGL_MESA_platform_surfaceless
00:00:00.017 [DEBUG] [wlr] [render/egl.c:469] Using EGL device /dev/dri/card0
libEGL warning: egl: failed to create dri2 screen
00:00:00.109 [ERROR] [wlr] [EGL] command: eglInitialize, error: EGL_NOT_INITIALIZED (0x3001), message: "DRI2: failed to create screen"
libEGL warning: NEEDS EXTENSION: falling back to kms_swrast

00:00:00.120 [DEBUG] [wlr] [render/egl.c:185] EGL DMA-BUF format modifiers unsupported
00:00:00.141 [DEBUG] [wlr] [render/egl.c:518] Using EGL_PLATFORM_DEVICE_EXT
00:00:00.141 [INFO] [wlr] [render/gles2/renderer.c:718] Creating GLES2 renderer
00:00:00.141 [INFO] [wlr] [render/gles2/renderer.c:719] Using OpenGL ES 3.2 Mesa 23.1.2
00:00:00.141 [INFO] [wlr] [render/gles2/renderer.c:720] GL vendor: Mesa
00:00:00.141 [INFO] [wlr] [render/gles2/renderer.c:721] GL renderer: llvmpipe (LLVM 16.0.5, 128 bits)

00:00:00.160 [DEBUG] [wlr] [render/allocator/gbm.c:210] Created GBM allocator with backend drm
00:00:00.160 [DEBUG] [wlr] [render/allocator/gbm.c:213] Using DRM node /dev/dri/card0
00:00:00.160 [DEBUG] [wlr] [types/wlr_idle.c:246] idle manager created
00:00:00.161 [DEBUG] [wlr] [types/wlr_drm_lease_v1.c:638] Creating wlr_drm_lease_device_v1 for /dev/dri/card0
00:00:00.161 [INFO] [wlr] [backend/headless/backend.c:68] Creating headless backend

Weston boot log:
Date: 2023-06-21 CST
[19:09:34.483] weston 12.0.1
               https://wayland.freedesktop.org
               Bug reports to: https://gitlab.freedesktop.org/wayland/weston/issues/
               Build: 12.0.1
[19:09:34.483] Command line: weston --log=/tmp/weston.log
[19:09:34.483] OS: Linux, 6.3.7, #1 SMP PREEMPT_DYNAMIC Tue Jun 13 17:47:41 CST 2023, x86_64
[19:09:34.483] Flight recorder: enabled
[19:09:34.483] Starting with no config file.
[19:09:34.484] Output repaint window is 7 ms maximum.
[19:09:34.484] Loading module '/usr/lib/libweston-12/drm-backend.so'
[19:09:34.485] initializing drm backend
[19:09:34.485] Trying libseat launcher...
[19:09:34.487] [libseat/libseat.c:73] Seat opened with backend 'logind'
[19:09:34.487] libseat: session control granted
[19:09:34.488] using /dev/dri/card0
[19:09:34.488] DRM: supports atomic modesetting
[19:09:34.488] DRM: does not support GBM modifiers
[19:09:34.488] DRM: supports picture aspect ratio
[19:09:34.506] Loading module '/usr/lib/libweston-12/gl-renderer.so'
[19:09:34.508] warning: failed to query rendering device from EGL
[19:09:34.508] EGL version: 1.5
[19:09:34.508] EGL vendor: Mesa Project
[19:09:34.508] EGL client APIs: OpenGL OpenGL_ES
[19:09:34.508] warning: Disabling render GPU timeline and explicit synchronization due to missing EGL_ANDROID_native_fence_sync extension
[19:09:34.508] EGL features:
               EGL Wayland extension: no
               context priority: no
               buffer age: yes
               partial update: no
               swap buffers with damage: no
               configless context: yes
               surfaceless context: yes
               dmabuf support: modifiers
[19:09:34.512] GL version: OpenGL ES 3.2 Mesa 23.1.2
[19:09:34.512] GLSL version: OpenGL ES GLSL ES 3.20
[19:09:34.512] GL vendor: Mesa
[19:09:34.512] GL renderer: llvmpipe (LLVM 16.0.5, 128 bits)
[19:09:34.517] GL ES 3.2 - renderer features:
               read-back format: ARGB8888
               glReadPixels supports y-flip: yes
               wl_shm 10 bpc formats: yes
               wl_shm 16 bpc formats: yes
               wl_shm half-float formats: yes
               internal R and RG formats: yes
               OES_EGL_image_external: yes
[19:09:34.517] Using GL renderer

### Please read the following before submitting:
- Please do NOT submit bug reports for questions. Ask questions on IRC at #sway on Libera Chat.
- Proprietary graphics drivers, including nvidia, are not supported. Please use the open source equivalents, such as nouveau, if you would like to use Sway.
- Please do NOT submit issues for information from the github wiki. The github wiki is community maintained and therefore may contain outdated information, scripts that don't work or obsolete workarounds.
  If you fix a script or find outdated information, don't hesitate to adjust the wiki page.

### Please fill out the following:
- **Sway Version:**
  - `swaymsg -t get_version` or `sway -v`

- **Debug Log:**
  - Run `sway -d 2> ~/sway.log` from a TTY and upload it to a pastebin, such as gist.github.com.
  - This will record information about sway's activity. Please try to keep the reproduction as brief as possible and exit sway.
  - Attach the **full** file, do not truncate it.

- **Configuration File:**
  - Please try to produce with the default configuration.
  - If you cannot reproduce with the default configuration, please try to find the minimal configuration to reproduce.
  - Upload the config to a pastebin such as gist.github.com.

- **Stack Trace:**
  - This is only needed if sway crashes.
  - If you use systemd, you should be able to open the coredump of the most recent crash with gdb with
    `coredumpctl gdb sway` and then `bt full` to obtain the stack trace.
  - If the lines mentioning sway or wlroots have `??` for the location, your binaries were built without debug symbols. Please compile both sway and wlroots from source and try to reproduce.

- **Description:**
  - The steps you took in plain English to reproduce the problem.

**Comment by emersion** [comment]
You're missing some GPU drivers.

```
libEGL warning: egl: failed to create dri2 screen
00:00:00.109 [ERROR] [wlr] [EGL] command: eglInitialize, error: EGL_NOT_INITIALIZED (0x3001), message: "DRI2: failed to create screen"
libEGL warning: NEEDS EXTENSION: falling back to kms_swrast
```
